### PR TITLE
feat: improve permission block display with structured details

### PR DESCRIPTION
**Pull request by FridayLiu**
## Problem

Permission approval prompts had several UX issues:

1. **Garbled tool name** — Machine-generated `toolCallId` (e.g. `toolu_vrtx_01Ajk...`) shown as a badge
2. **Raw JSON details** — Request context displayed as `JSON.stringify()` output
3. **No visual differentiation** — All permission types looked identical
4. **Duplicate information** — File paths repeated across title, targets, and details
5. **Always-approve broken** — Copilot "Always Allow" leaked across sessions due to global `Set`

### Architecture: Adapter-layer normalization

Following the engine-agnostic frontend rule, each adapter now populates a structured `details: PermissionDetail[]` field on `UnifiedPermission`. The frontend renders these directly — no field-name guessing.

```typescript
interface PermissionDetail {
  label: string;   // Already resolved text (e.g. "Command", "URL")
  value: string;
  mono?: boolean;  // Render in monospace
}
```

### Per-engine details extraction

| Engine | Extracted details |
|--------|-------------------|
| **Copilot** | All 8 SDK kinds via `PermissionPromptRequest`: `fullCommandText` (commands), `fileName`/`diff` (write), `path` (read), `url`, `toolTitle`/`serverName` (mcp), `subject`/`fact` (memory), `toolName`/`toolDescription` (custom-tool), `hookMessage` (hook), `paths` (path) |
| **OpenCode** | `command`, `url`, `path`, `reason` from dynamic `metadata` object |
| **Claude** | Auto-approve all via `canUseTool` (TODO: populate when permission mode changes) |

### Copilot SDK 0.3.0 upgrade

Upgraded `@github/copilot-sdk` from exact `0.2.0` to exact `0.3.0`:

- **`PermissionPromptRequest`**: New UI-oriented type providing `intention`, `canOfferSessionApproval`, `diff`, `fileName`, `commandIdentifiers`, etc. The adapter now consumes this from `permission.requested` session events for richer permission prompts.
- **Permission result contract**: Adapted to new result types (`approve-once`, `reject`, `user-not-available`, `approve-for-session`) replacing the old `approved`/`denied-*` values.
- **Native session approval**: When `canOfferSessionApproval` is true and a valid approval payload can be constructed, "Always Allow" now sends `approve-for-session` with typed approval (commands, write, read, mcp, memory, custom-tool) instead of client-side tracking.
- **Conditional "Always Allow"**: The button is only shown when the SDK indicates session-wide approval is available.
- **Fallback compatibility**: When `promptRequest` is absent (older CLI protocol), falls back to constructing details from `permissionRequest` fields.

### UI improvements

- Kind-specific SVG icons extracted to `src/components/icons/permission-icons.tsx`
- Tool-to-label mapping via lookup tables instead of hardcoded switch chains
- Removed redundant `getPermissionTargets` frontend extraction (details[] is now canonical)
- Removed opaque `toolCallId` badge from question prompts
- `ContentDiff` now accepts a `fallback` prop, removing `parsePatch` coupling from permission component
- Cleaned up unused CSS classes, i18n keys (`targets`, `relatedTool`), and related tests

## Changes

| File | What changed |
|------|-------------|
| `package.json` / `bun.lock` | `@github/copilot-sdk` `0.2.0` → `0.3.0` |
| `electron/main/engines/copilot/index.ts` | Consume `PermissionPromptRequest` from session events; adapt to new permission result contract; conditional "Always Allow" with native `approve-for-session`; extracted helper functions |
| `electron/main/engines/claude/index.ts` | Added TODO for future permission detail support |
| `src/components/InputAreaPermission.tsx` | Render only `details[]`; use extracted icon components and lookup tables |
| `src/components/InputAreaQuestion.tsx` | Removed `toolCallId` badge display |
| `src/components/input-area-context.ts` | Removed `getPermissionTargets` and related helpers |
| `src/components/icons/permission-icons.tsx` | New: extracted SVG icon components |
| `src/components/share/content-diff.tsx` | Added `fallback` prop for graceful parse failures |
| `src/locales/{en,zh,ru}.ts` | Removed unused `targets` and `relatedTool` keys |
| `tests/unit/` | Updated permission result expectations; added `PermissionPromptRequest` and session approval tests |

## Testing

- `npm run typecheck` ✅
- `bun run test:unit` — 68 files, 2311 tests ✅
- `npm run build` ✅
- `npm run lint` — 0 errors ✅

### CODE REVIEW

File: src/types/unified.ts [hunk 1]
```diff
@@ -417,19 +417,36 @@ export interface PermissionOption {
...
   kind: "read" | "edit" | "other";
   /** Diff preview for write operations */
   diff?: string;
-  /** Raw input for context */
+  /**
+   * Structured display details, populated by the adapter layer.
+   * Each item is a label/value pair ready for rendering — no frontend parsing needed.
+   */
+  details?: PermissionDetail[];
+  /** Raw input for context (legacy fallback) */
```
[the remark this answers is not in the document]
Comment (FridayLiu) [reply src/types/unified.ts:456]: PR description has been updated to accurately reflect the adapter + type changes. No longer claims pure UI change.

File: electron/main/engines/opencode/index.ts [hunk 2]
```diff
@@ -571,17 +571,38 @@ export class OpenCodeAdapter extends EngineAdapter {
...
+      title: data.title ?? data.permission ?? "Permission request",
       kind: "edit",
+      details,
       rawInput: data.metadata,
       options,
-      permission: data.type,
-      patterns: data.pattern ? (Array.isArray(data.pattern) ? data.pattern : [data.pattern]) : [],
+      permission: data.permission,
+      patterns: Array.isArray(data.patterns) ? data.patterns : [],
     };
```
[the remark this answers is not in the document]
Comment (FridayLiu) [reply electron/main/engines/opencode/index.ts]: Fixed in 74e35d7. Now derives kind from data.permission: bash/shell -> edit, read -> read, others -> other.

File: src/components/InputAreaPermission.tsx [hunk 3]
```diff
@@ -57,10 +95,56 @@ export function InputAreaPermission(props: InputAreaPermissionProps) {
         <span class={styles.headerLabel}>{t().permission.waitingApproval}</span>
       </div>
 
+      <div class={styles.meta}>
+        <span class={styles.metaBadge}>
+          <svg class={styles.kindIcon} xmlns="http://www.w3.org/2000/svg" width="12" height="12" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
+            <path d={getKindIcon(props.permission.kind, props.permission.toolName)} />
+          </svg>
+          {kindLabel()}
+        </span>
```
[the remark this answers is not in the document]
Comment (FridayLiu) [reply src/components/InputAreaPermission.tsx:83]: Intentional removal. The toolCallId is a machine-generated ID (e.g. toolu_vrtx_01Ajk...) that provides no value to users. Removed the unused toolCall i18n key in 74e35d7.

File: src/components/InputAreaPermission.tsx [hunk 4]
```diff
@@ -57,10 +95,56 @@ export function InputAreaPermission(props: InputAreaPermissionProps) {
...
+        </div>
+      </Show>
+
+      <Show when={props.permission.diff}>
+        <div class={styles.contextSection}>
+          <div class={styles.contextLabel}>{t().permission.diffPreview}</div>
+          <div class={styles.diffContainer}>
+            <ContentDiff diff={props.permission.diff!} />
+          </div>
+        </div>
```
[the remark this answers is not in the document]
Comment (FridayLiu) [reply src/components/InputAreaPermission.tsx]: Fixed in 74e35d7. Now pre-checks with parsePatch() whether the diff is parseable. If it yields no hunks, falls back to plain preformatted text.

File: src/components/input-area-context.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+  if (requestPreview) {
+    return {
+      type: "request",
+      content: requestPreview,
+    };
+  }
+
+  return null;
+}
+
```
[the remark this answers is not in the document]
Comment (FridayLiu) [reply src/components/input-area-context.ts]: Removed in 74e35d7. getPermissionPreview, buildRequestPreview, formatCommandPreview, compactPreview, PREVIEW_OMIT_KEYS, and the associated test case have all been deleted. Permission details are now fully adapter-provided via details[], no frontend guessing needed.

File: src/types/unified.ts [hunk 6]
```diff
@@ -417,19 +417,36 @@ export interface PermissionOption {
...
   /** Operation kind */
   kind: "read" | "edit" | "other";
   /** Diff preview for write operations */
   diff?: string;
-  /** Raw input for context */
+  /**
+   * Structured display details, populated by the adapter layer.
+   * Each item is a label/value pair ready for rendering — no frontend parsing needed.
+   */
+  details?: PermissionDetail[];
```
Comment (realDuang) [inline src/types/unified.ts:455]: Copilot, OpenCode, and Codex adapters all populate `details` and `toolName` now, but the Claude adapter (`electron/main/engines/claude/index.ts`, around its `handlePermissionRequest` / `canUseTool` flow) is untouched. It will emit `UnifiedPermission` with `details: undefined` and `toolName: undefined`, so Claude users will see no structured details and only the generic gear icon.

Worth adding Claude coverage in this PR — or at minimum document it as a known gap if it's intentionally deferred.

File: src/components/input-area-context.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+export interface QuestionContextInfo {
+  toolCallId?: string;
+  isMultiQuestion: boolean;
+  current: number;
+  total: number;
+}
+
+export function getPermissionTargets(
```
Comment (realDuang) [inline src/components/input-area-context.ts]: This function probes `rawInput` and `metadata` looking for path-like keys to extract targets — but the new `details: PermissionDetail[]` field populated by adapters already carries the same information (commands, paths, URLs) in a structured form.

Having both means the frontend is doing adapter-level extraction again, which is the pattern this PR set out to eliminate. If `details[]` is the canonical structured representation, consider either:
- Removing `getPermissionTargets` and rendering only `details[]`, or
- Having adapters populate a dedicated `targets?: string[]` on `UnifiedPermission` instead of the frontend guessing from `rawInput` field names.

As-is, you could end up showing the same file path in both the "Targets" chip list and the "Path" detail row.

File: src/components/InputAreaPermission.tsx [hunk 8]
```diff
@@ -46,6 +64,27 @@ export function InputAreaPermission(props: InputAreaPermissionProps) {
         ];
   };
 
+  const kindLabel = () => {
```
Comment (realDuang) [inline src/components/InputAreaPermission.tsx:54]: This maps `toolName` strings to i18n labels via hardcoded comparisons (`"web_fetch"`, `"shell"`, etc.). If a new engine introduces a tool name not listed here (e.g. `"container_exec"`), it silently falls through to `kindOther`.

Consider a lookup table approach — something like `const TOOL_KIND_MAP: Record<string, () => string>` — so it's easier to extend, and the fallback is explicit rather than implicit via `switch/default`.

File: src/components/InputAreaPermission.tsx [hunk 9]
```diff
@@ -1,17 +1,38 @@
...
 
 interface InputAreaPermissionProps {
   permission: UnifiedPermission;
   onRespond: (sessionID: string, permissionID: string, reply: string) => void;
 }
 
+const MAX_VISIBLE_TARGETS = 4;
+
+/** SVG icon paths keyed by permission kind + toolName */
+const KIND_ICONS: Record<string, string> = {
```
Comment (realDuang) [inline src/components/InputAreaPermission.tsx]: Long SVG path-data strings inlined in a component file are hard to maintain (especially the gear fallback at line 29, which is ~500 chars). The project already has `src/components/icons/` for this purpose — consider extracting these into that directory, or at least pulling the path constants into a separate file.

File: electron/main/engines/copilot/index.ts [hunk 10]
```diff
@@ -1316,20 +1325,90 @@ export class CopilotSdkAdapter extends EngineAdapter {
...
+    const toolName = req.kind === "url" ? "web_fetch" : req.kind === "shell" ? "shell" : req.kind === "read" ? "read" : req.kind === "write" ? "edit" : req.kind === "mcp" ? "mcp" : undefined;
     const options: PermissionOption[] = [
       { id: "allow_once", label: "Allow Once", type: "allow_once" },
       { id: "allow_always", label: "Always Allow", type: "allow_always" },
       { id: "reject_once", label: "Deny", type: "reject_once" },
     ];
 
-    const permission: UnifiedPermission = { id: permissionId, sessionId, engineType: this.engineType, toolCallId: req.toolCallId, title: req.title || `${req.kind} permission requested`, kind, diff: req.diff, rawInput: { ...req }, options };
+    const rawReq = { ...req };
+    const intention = typeof rawReq.intention === "string" && rawReq.intention.trim() ? rawReq.intention.trim() : "";
```
Comment (realDuang) [inline electron/main/engines/copilot/index.ts]: Where does `intention` come from? The `PermissionRequest` interface (line ~78) doesn't declare it, and the Copilot SDK's `PermissionRequestEvent` type doesn't appear to have this field either. If it's a real SDK field, it should be added to the interface; if it's speculative/future-proofing, this code will never trigger and adds dead complexity.

File: src/components/InputAreaQuestion.tsx [hunk 11]
```diff
@@ -147,6 +153,20 @@ export function InputAreaQuestion(props: InputAreaQuestionProps) {
         </Show>
       </div>
 
+      <Show when={questionContext().toolCallId || questionContext().isMultiQuestion}>
+        <div class={styles.meta}>
+          <Show when={questionContext().toolCallId}>
+            <span class={styles.metaBadge}>
+              <span>{t().question.relatedTool}</span>
+              <span class={styles.metaMono}>{questionContext().toolCallId}</span>
```
Comment (realDuang) [inline src/components/InputAreaQuestion.tsx]: One of this PR's motivations is removing the garbled `toolCallId` badge from permission prompts (since machine-generated IDs like `toolu_vrtx_01Ajk...` provide no user value). But here a `toolCallId` is being *added* to the question prompt UI. The same readability concern applies — these IDs are opaque to users.

If the intent is to help users correlate questions with tool calls, consider showing the tool *name* instead (which the question's parent context might provide), or drop this badge entirely for consistency.

File: src/components/InputAreaPermission.tsx [hunk 12]
```diff
@@ -1,17 +1,38 @@
 import { For, Show } from "solid-js";
-import type { UnifiedPermission } from "../types/unified";
+import { parsePatch } from "diff";
```
Comment (realDuang) [inline src/components/InputAreaPermission.tsx]: Nit: `parsePatch` is imported here only to check whether a diff string is parseable (line ~136). `ContentDiff` already handles rendering and could absorb this check internally — e.g. it could accept a `fallback` prop or handle parse failures gracefully. That would avoid coupling the permission component to the `diff` library's API surface.

**Review by FridayLiu — COMMENTED** [review]
## Security Review — PR #121

### Summary

This PR upgrades Copilot SDK permission result types, adds structured permission detail display, scopes "Always Allow" per-session, and adds Claude autopilot mode with permission prompts for non-autopilot modes. I reviewed all changed files from a security perspective. **No blocking security issues found.** One low-severity finding and a few informational notes below.

---

### 1. Permission Result Mapping (Copilot) ✅

The old `approved`/`denied-*` values are correctly replaced:

- `approvePermissionOnce()` → `{ kind: "approve-once" }` — correct
- `rejectPermission()` → `{ kind: "reject" }` — correct
- `userNotAvailablePermission()` → `{ kind: "user-not-available" }` — correct, used for disconnects/cleanup
- `approve-for-session` is only emitted when `pending.sessionApproval` is present AND user explicitly chose `allow_always` — no path where a malformed result could accidentally grant session-wide approval

**Fail-safe behavior**: When `allow_always` is chosen but no `sessionApproval` object can be constructed (e.g., shell commands without identifiable `commandIdentifiers`), it falls back to the per-session `allowedAlwaysKinds` map and returns `approve-once` to the SDK. This is a safe degradation.

### 2. `approve-for-session` Scoping Fix ✅

**Old code**: `allowedAlwaysKinds` was a flat `Set<string>` on the adapter instance — global across all sessions. A user approving `shell` in session A would auto-approve `shell` in session B.

**New code**: `allowedAlwaysKinds` is now `Map<string, Set<string>>` keyed by `sessionId`. The `deleteSession()` method correctly calls `this.allowedAlwaysKinds.delete(sessionId)`. This properly scopes the fallback "Always Allow" per session.

### 3. TOCTOU / Race Conditions ✅

The `queueMicrotask` in `handlePermissionRequest` is used to let the `permission.requested` event arrive before building the UI prompt. This is a producer-consumer pattern with:
- `rememberPermissionPromptContext` (event handler) pushes to a queue
- `consumePermissionPromptContext` (microtask) shifts from the queue

The key is composed of `sessionId:toolCallId` (or `sessionId:kind` as fallback). Since the SDK fires `permission.requested` synchronously before calling `onPermissionRequest`, the microtask ordering is guaranteed to see the context. No race condition here.

### 4. XSS in Permission Details ✅

`InputAreaPermission.tsx` renders `detail.value` and `detail.label` via SolidJS text interpolation (`{detail.value}`), which automatically escapes HTML. No `innerHTML` or `dangerouslySetInnerHTML` is used. The `ContentDiff` component also uses safe rendering (text nodes, no innerHTML). **No XSS risk.**

The `title` field and SVG `path d=` attribute use `getKindIconPath()` which only returns values from a hardcoded constant map — not from engine input.

### 5. Claude Auto-Approve / Autopilot Mode ⚠️ (Informational)

When Claude is in `autopilot` mode, `canUseTool` returns `{ behavior: "allow" }` for **all tools** except `ExitPlanMode`. This means shell commands, file writes, web fetches — everything is auto-approved with zero user interaction.

**This is by design** (the mode is explicitly labeled "Fully autonomous, skip all permission prompts"), and critically:
- The default mode is `"default"`, not `"autopilot"` — users must opt in
- The old code was *worse*: it auto-approved everything unconditionally (`return { behavior: "allow", updatedInput: input }`) regardless of mode. This PR actually **adds** permission prompts for default/acceptEdits modes, which is a net security improvement.

**Recommendation** (non-blocking): Consider adding a one-time confirmation dialog when switching to autopilot mode, especially for remote sessions where the security implications are higher.

### 6. `permissionPromptContexts` Memory Leak (Low Severity)

`permissionPromptContexts` (`electron/main/engines/copilot/index.ts:215`) is never cleaned up in `deleteSession()` (line 436-458). If a `permission.requested` event fires but `onPermissionRequest` is never called (e.g., the session is deleted mid-flight), contexts will accumulate. The keys are `sessionId:toolCallId`, so they would not be reused across sessions, but the entries persist until the adapter is garbage collected.

**Impact**: Very low — only leaks small objects, and only in edge cases where sessions are deleted during active permission flows. But for correctness, `deleteSession` should clear entries with matching session prefix.

### 7. Dependency Pin ✅

`@github/copilot-sdk` changed from `^0.3.0` to `0.3.0` (exact pin). This is good practice for an SDK with breaking permission API changes — prevents unexpected upgrades from breaking the permission contract.

---

### Verdict: **No blocking security issues.** The permission result mapping is correct and complete. The session scoping fix for `allowedAlwaysKinds` closes a real cross-session permission leak. The Claude changes are a net security improvement. XSS is not a concern. The `permissionPromptContexts` cleanup gap is low-severity and non-blocking.

**Review by FridayLiu — COMMENTED** [review]
## Test Coverage & Quality Review

### ✅ What's Good

1. **Copilot permission tests updated correctly** — The `PendingPermission` shape change (`sdkKind`, `sessionApproval` instead of `rawInput.kind`) is consistently reflected across all existing tests in `copilot/index.test.ts`.

2. **New `PermissionPromptRequest` integration test** (line ~1112) effectively validates the enriched permission flow: title from `intention`, structured `details`, `diff` passthrough, and `approve-for-session` result when `canOfferSessionApproval: true`.

3. **Session-scoped `allowedAlwaysKinds`** — The test at line ~1079 correctly validates the Map-of-Sets migration from the old global Set.

4. **`user-not-available` migration** — Clean update from `{ kind: "reject" }` to `{ kind: "user-not-available" }` in the stop/cleanup test.

5. **OpenCode test fixes** (pattern → array, permission field rename) look correct and match source changes.

6. **i18n keys are in sync** — All three locale files (`en.ts`, `zh.ts`, `ru.ts`) add the same 7 permission keys (`kindRead`, `kindEdit`, `kindOther`, `kindShell`, `kindUrlAccess`, `kindWebSearch`, `diffPreview`) and the `question.progress` key. ✅

7. **`getPermissionTargets` removal** — Confirmed fully cleaned up; no remaining references anywhere in the codebase. ✅

---

### ⚠️ Coverage Gaps

#### 1. **Copilot: Only 2 of ~9 SDK permission kinds tested**

The new `buildPermissionDetails()` has switch branches for `commands/shell`, `write`, `read`, `url`, `mcp`, `path`, `memory`, `custom-tool`, and `hook`. Only `write` and `read` have integration tests (lines ~1112, ~1149). Missing coverage for:
- `commands`/`shell` — `fullCommandText`, `warning`, `commandIdentifiers` extraction
- `url` — URL detail extraction
- `mcp` — `serverName`/`toolName` detail + session approval
- `path` — `accessKind` mapping, paths array
- `memory` — action/subject/fact/direction/reason/citations details
- `custom-tool` — toolName/toolDescription details
- `hook` — toolName/hookMessage details

These are distinct code paths with unique field extraction logic and would benefit from parameterized tests.

#### 2. **Copilot: `buildSessionApproval()` only tested for `write` kind**

The test at line ~1136 validates `{ kind: "write" }` approval. Missing tests for:
- `commands` approval with `commandIdentifiers`
- `mcp` approval with `serverName`/`toolName`
- `custom-tool` approval with `toolName`
- Fallback `undefined` when required fields are missing (e.g., `mcp` without `serverName`)

#### 3. **Copilot: No test for fallback when `promptRequest` is absent**

When `permission.requested` event fires without `promptRequest`, the code falls back to `permissionRequest` as the source. The existing test at line ~1098 exercises this path implicitly (no `handleSessionEvent` call), but it would be clearer to have an explicit test that fires `permission.requested` with `promptRequest: undefined` to verify the fallback chain.

#### 4. **Codex converters: `details` and `toolName` not asserted**

The PR adds `details: PermissionDetail[]` and `toolName` to all 6 return paths in `convertApprovalToPermission`. The existing ~40 codex converter tests use `toMatchObject` and none assert `details` or `toolName`. These new fields pass silently even if broken.

#### 5. **Claude adapter: No tests for `handleToolPermission` or autopilot mode**

The PR adds a significant new `handleToolPermission()` method and autopilot mode logic in the Claude adapter, but `tests/unit/electron/engines/claude/` has no coverage for these.

#### 6. **`content-diff.tsx` fallback: No test**

The new fallback rendering (lines 67-72 in content-diff.tsx) — when `lines()` is empty but `props.diff` is truthy, render as `<pre>` — has no test coverage.

---

### 📝 Test Quality Notes

#### New `permission-icons.test.ts` — Adequate but minimal
The 3 test cases cover the happy paths well (tool match, kind fallback, unknown fallback). Consider adding: `toolName` that exists but `kind` is also valid (priority test), and `undefined`/empty string toolName.

#### New `input-area-context.test.ts` — Only tests multi-question
Missing single-question case (`questions.length === 1` → `isMultiQuestion: false`), empty questions edge case, and `pageIndex` out of bounds (e.g., `pageIndex >= questions.length` which triggers the `Math.min` clamping logic at line 14).

#### Copilot test: `await Promise.resolve()` for microtask flush
The new tests correctly add `await Promise.resolve()` after `handlePermissionRequest` to flush the `queueMicrotask` delay — good attention to the implementation change. However, this couples tests to the implementation detail of using `queueMicrotask`. A helper like `flushMicrotasks()` would make intent clearer.

---

### Summary

The PR tests cover the critical happy path for the new permission prompt enrichment, but leave significant gaps in edge cases (6 of 9 permission kinds untested), the codex adapter changes, and the Claude adapter changes. The new component tests are a good start but could be more thorough.
